test/default.nix

```diff
@@ -181,12 +181,15 @@ let
     stack-source-repo = callTest ./stack-source-repo {};
     extra-hackage = callTest ./extra-hackage { inherit compiler-nix-name; };
     compiler-nix-name = callTest ./compiler-nix-name {};
-    hls-cabal = callTest ./haskell-language-server/cabal.nix { inherit compiler-nix-name; };
-    hls-stack = callTest ./haskell-language-server/stack.nix { inherit compiler-nix-name; };
     cabal-hpack = callTest ./cabal-hpack { inherit util compiler-nix-name; };
-    index-state = callTest ./index-state { inherit compiler-nix-name; };
 
     unit = unitTests;
+  } // lib.optionalAttrs (builtins.compareVersions pkgs.buildPackages.haskell-nix.compiler.${compiler-nix-name}.version "9.0" < 0) {
```

Comment:
There's the `disabled` meta property on the tests, which I think we could use for this.

Reply:
TBH there's a bunch of this sort of conditional in here that we could probably push into there...

Reply:
I'm not too keen on disabling these and I am hoping we can mark them as `broken` instead.  I'll try to do that in the meta property. It is going to be tricky though.

The problem, with `hls-cabal` for instance, is that it fails in `cabal configure` for ghc 9.0.1 (one day hackage will be updated and it will start working).  The error shows up at hydra evaluation time so there is no good way to ignore it.

I tried adding a hackage quirk as a work around:

```
allow-newer: *:base, *:ghc-prim, *:template-haskell, *:ghc-boot-th, *:ghc
```

This is probably a really bad idea as it fixes `cabal configure` (and makes the hydra eval work), but still hls does not build.  Even if it did we would have no way of knowing when it is actually fixed in hackage or not.

My plan is:

* Make it so `cabal configure` errors are not eval time errors see #1060.

* Exclude derivations marked as `broken` from the `required-*` aggregates.

* Include derivations marked as `broken` in the `release.nix` (so each one has its own entry on hydra and we can see when they are fixed).

Reply:
> The error shows up at hydra evaluation time so there is no good way to ignore it.

Is this true? Don't we get saved by laziness? That is, provided we don't force the "definition" of the test and just check on the meta attribute to see whether to include it in our CI jobset, we should be fine?

> I'm not too keen on disabling these and I am hoping we can mark them as broken instead.

`disabled` is just our own version of `broken` that doesn't break evaluation merely by being present, and just causes it to be filtered out of the jobset, see https://github.com/input-output-hk/haskell.nix/pull/976

Reply:
> Include derivations marked as broken in the release.nix (so each one has its own entry on hydra and we can see when they are fixed).

Maybe we need to chat about the policy here. What we have today:
- The required set is not all the jobs
- `disabled` derivations *don't* appear in the jobset

The nice thing about this is that it means that we can say that the master jobset should always be completely green, even if it's not all in the required job. We had people being confused before by "known failures" in the main jobset even if they weren't required.

Reply:
> Is this true? Don't we get saved by laziness?

The hydra job json that eval generated includes the `.drv`.  On master of haskell.nix the following will fail to create a derivation due to the imposable constraint:

```
nix-instantiate -E '(import ./. {}).pkgs-unstable.haskell-nix.tool "ghc8104" "hello" { cabalProject = "packages: .\nconstraints: hello<0"; }'
```

The same thing on the #1060 branch returns a derivation.  When you `nix-build` the derivation though it will display the `cabal configure` output and fail.

> The nice thing about this is that it means that we can say that the master jobset should always be completely green, even if it's not all in the required job. We had people being confused before by "known failures" in the main jobset even if they weren't required.

I would like to know when a "known failure" (due to something in hackage) turns green as a result of the daily hackage update.  I'm not sure how we can do that if we don't allow them to show up as red.  I'll see if I can set it up to include `disabled` on `ci.zw3rk.com` and leave them out on `hydra.iohk.io`.

Reply:
> The hydra job json that eval generated includes the .drv. On master of haskell.nix the following will fail to create a derivation due to the imposable constraint:

Sure, but the disabled jobs don't go into the hydra job *at all*. So we only inspect `meta.disabled` and decide not to include it, so I don't think the eval should fail. Maybe I just need to try this to prove it :p

> I would like to know when a "known failure" (due to something in hackage) turns green as a result of the daily hackage update.

Yeah, I think we just don't have an ideal way of dealing with "known failures" in Hydra:
- If we put them in the jobset, the jobset looks more broken than it is
- If we leave them out we don't notice when they start working.

Ideally they'd be marked specially and we'd get some kind of notification when they change state!

Reply:
> Sure, but the disabled jobs don't go into the hydra job at all. So we only inspect meta.disabled and decide not to include it, so I don't think the eval should fail. Maybe I just need to try this to prove it :p

Sorry I must have misunderstood what you were saying.  Disabled tests do not cause eval issues.

Reply:
Oh I see, you just plain don't want to disable it because it doesn't show up then. Fair enough, I guess.

Reply:
Personally I'd just disable it and audit the disabled conditions every now and again, but that is somewhat annoying.

Reply:
And in this case, setting `disabled` has the same effect as these conditionals, so it's no worse and at least it's in one place.

Reply:
Here is my typical process for getting hydra to build haskell.nix when there are eval issues:

* Delve into the eval log on hydra (checking that it is a real eval failure not just OOM or GC issue).

* Try to reproduce it locally using `nix repl release.nix`, `nix-build` or `hydra-eval-jobs`.

* Fix it if possible or disable the test so other eval errors show up.  It is difficult to be sure all the jobs are running unless you see `Last Checked: ... no errors`.

* Repeat until eval works.

* Fix any jobs that are failing. <- This is where I am now with this PR.

* Go back to see which of the disable jobs can be enabled and fixed now the other issues are resolved.

I'm really hoping #1060 will alleviate some of the eval time issues.  Unfortunately some of the tests are written in a way that will not take advantage straight away, but I have updated some of the more problematic ones (the HLS test is the worst).

Reply:
Right, tests whose eval state *changes* are very annoying indeed. Time to go back to Hercules? :trollface: